services/logs/logs.go

```diff
@@ -164,54 +172,84 @@ func New(pmmVersion string, consul *consul.Client, rds *rds.Service, logs []Log)
 	return l
 }
 
+// Files returns list of logs and their content.
+func (l *Logs) Files(ctx context.Context) []File {
+	files := make([]File, 0, len(l.logs))
+
+	for _, log := range l.logs {
+		var f File
+		f.Name, f.Data, f.Err = l.readLog(ctx, &log)
+		files = append(files, f)
+	}
+
+	var agents []reform.Struct
+	err := l.db.InTransaction(func(tx *reform.TX) error {
+		var node models.Node
+		err := tx.FindOneTo(&node, "type", models.PMMServerNodeType)
+		if err != nil {
+			return errors.Wrap(err, "failed to get PMM Server node")
+		}
+
+		agents, err = tx.FindAllFrom(models.AgentTable, "runs_on_node_id", node.ID)
+		return errors.Wrap(err, "failed to get agents running in PMM Server")
+	})
+	if err != nil {
+		logger.Get(ctx).WithField("component", "logs").Errorf("Failed to get RDS agents: %s.", err)
+	}
+
+	// collect names in set, not in slice,
+	// in case a single agent does several jobs and is returned several times
+	names := make(map[string]struct{}, len(agents))
+	for _, a := range agents {
+		names[a.(*models.Agent).NameForSupervisor()] = struct{}{}
+	}
+
+	for name := range names {
+		log := Log{
+			FilePath: logsRootDir + name + ".log",
+			UnitName: name,
+		}
+		var f File
+		f.Name, f.Data, f.Err = l.readLog(ctx, &log)
+		files = append(files, f)
+	}
+
+	sort.Slice(files, func(i, j int) bool { return files[i].Name < files[j].Name })
+	return files
+}
+
 // Zip creates .zip archive with all logs.
 func (l *Logs) Zip(ctx context.Context, w io.Writer) error {
 	zw := zip.NewWriter(w)
 	now := time.Now().UTC()
-	for _, log := range l.logs {
-		name, content, err := l.readLog(ctx, &log)
-		if name == "" {
-			continue
-		}
-
-		if err != nil {
-			logger.Get(ctx).WithField("component", "logs").Error(err)
+	for _, file := range l.Files(ctx) {
+		if file.Err != nil {
+			logger.Get(ctx).WithField("component", "logs").Error(file.Err)
 
 			// do not let a single error break the whole archive
-			if len(content) > 0 {
-				content = append(content, "\n\n"...)
+			if len(file.Data) > 0 {
+				file.Data = append(file.Data, "\n\n"...)
 			}
-			content = append(content, []byte(err.Error())...)
+			file.Data = append(file.Data, []byte(file.Err.Error())...)
+		}
+
+		if file.Name == "" {
```

Comment:
maybe it should be placed before `if file.Err != nil {` line ?

Reply:
because if file.Name == "" above operations don't make sense.

Reply:
Actually, this line no longer has any sense. :)

Reply:
Well, this is one corner case when it is needed…